subsys/bluetooth/host/gatt.c

```diff
@@ -5696,3 +5697,155 @@ void bt_gatt_disconnected(struct bt_conn *conn)
+	const struct bt_gatt_service_val *service_val;
+	uint8_t *data_buf = data_state->data_buf;
+	struct bt_data *data = data_state->data;
+	const uint8_t data_type = data->type;
+	const struct bt_uuid *service_uuid;
+	uint8_t service_uuid_type;
+
+	/* skip if attribute is not a service */
+	if (bt_uuid_cmp(attr->uuid, BT_UUID_GATT_PRIMARY) != 0 &&
+	    bt_uuid_cmp(attr->uuid, BT_UUID_GATT_SECONDARY) != 0) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	service_val = (void *)attr->user_data;
+	service_uuid = (void *)&service_val->uuid;
+	service_uuid_type = service_uuid->type;
+
+	if (service_uuid_type == BT_UUID_TYPE_16 &&
+	    (data_type == BT_DATA_UUID16_SOME ||
+	     data_type == BT_DATA_UUID16_ALL)) {
+		const struct bt_uuid_16 *uuid_16 = (void *)service_uuid;
+		const uint16_t uuid_val = BT_UUID_16(uuid_16)->val;
+		const size_t len = BT_UUID_SIZE_16;
+
+		/* We don't want to advertise some specific UUID16s */
+		for (size_t i = 0; i < ARRAY_SIZE(svc_uuid16_data_exclude_list); i++) {
+			if (uuid_val == svc_uuid16_data_exclude_list[i]) {
+				BT_DBG("Skipping UUID16 0x%04X", uuid_16->val);
+				return BT_GATT_ITER_CONTINUE;
+			}
+		}
+
+		if (data->data_len + len > data_buf_size) {
+			data_state->truncated = true;
+			return BT_GATT_ITER_STOP;
+		}
+
+		BT_DBG("Adding UUID16 0x%04X", uuid_val);
+		sys_put_le16(uuid_val, &data_buf[data->data_len]);
+		data->data_len += len;
+	} else if (service_uuid_type == BT_UUID_TYPE_32 &&
+		   (data_type == BT_DATA_UUID32_SOME ||
+		    data_type == BT_DATA_UUID32_ALL)) {
+		const struct bt_uuid_32 *uuid_32 = (void *)service_uuid;
+		const size_t len = BT_UUID_SIZE_32;
+
+		if (data->data_len + len > data_buf_size) {
+			data_state->truncated = true;
+			return BT_GATT_ITER_STOP;
+		}
+
+		BT_DBG("Adding UUID32 0x%08X", uuid_32->val);
+		sys_put_le32(uuid_32->val, &data_buf[data->data_len]);
+		data->data_len += len;
+	} else if (service_uuid_type == BT_UUID_TYPE_128 &&
+		   (data_type == BT_DATA_UUID128_SOME ||
+		    data_type == BT_DATA_UUID128_ALL)) {
+		const struct bt_uuid_128 *uuid_128 = (void *)service_uuid;
+		const size_t len = BT_UUID_SIZE_128;
+
+		if (data->data_len + len > data_buf_size) {
+			data_state->truncated = true;
+			return BT_GATT_ITER_STOP;
+		}
+
+		BT_DBG("Adding UUID128 %s", bt_hex(uuid_128->val, len));
+		(void)memcpy(&data_buf[data->data_len], uuid_128->val, len);
+		data->data_len += len;
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+int bt_gatt_get_svc_uuid_data(uint8_t type, struct bt_data *data,
+			      uint8_t data_buf[], size_t data_buf_size)
+{
+	struct svc_uuid_data_state data_state = {
+		.data = data,
+		.data_buf = data_buf,
+		.data_buf_size = data_buf_size
+	};
+
+	CHECKIF(!is_valid_adv_uuid_type(type)) {
+		BT_DBG("Invalid type: %u", type);
+		return -EINVAL;
+	}
+
+	CHECKIF(data == NULL) {
+		BT_DBG("data is NULL");
+		return -EINVAL;
+	}
+
+	CHECKIF(data_buf == NULL) {
+		BT_DBG("data_buf is NULL");
+		return -EINVAL;
+	}
+
+	data->type = type;
```

Comment:
This is a question, not a suggestion.. Should `data` be initialized / zeroed out here?

Reply:
Not sure I completely follow. 

`bt_data` is a representation of a LTV structure. Resetting it here would not make much sense as we are overwriting all 3 values in the struct already.